pageserver/src/layered_repository/timeline.rs

```diff
@@ -634,6 +645,56 @@ impl LayeredTimeline {
         Ok(())
     }
 
+    /// (Re-)calculate the logical size of the database at the latest LSN.
+    ///
+    /// This can be a slow operation.
+    pub fn init_logical_size(&self) -> Result<()> {
+        // Try a fast-path first:
+        // Copy logical size from ancestor timeline if there has been no changes on this
+        // branch, and no changes on the ancestor branch since the branch point.
+        if self.get_ancestor_lsn() == self.get_last_record_lsn() && self.ancestor_timeline.is_some()
+        {
+            let ancestor = self.get_ancestor_timeline()?;
+            let ancestor_logical_size = ancestor.get_current_logical_size();
+            // Check LSN after getting logical size to exclude race condition
+            // when ancestor timeline is concurrently updated
+            if ancestor_logical_size != 0 && ancestor.get_last_record_lsn() == self.ancestor_lsn {
+                self.current_logical_size
+                    .store(ancestor_logical_size as isize, AtomicOrdering::SeqCst);
+                debug!(
+                    "logical size copied from ancestor: {}",
+                    ancestor_logical_size
+                );
+                return Ok(());
+            }
+        }
+
+        // Have to calculate it the hard way
+        let last_lsn = self.get_last_record_lsn();
+        let logical_size = self.get_current_logical_size_non_incremental(last_lsn)?;
+        self.current_logical_size
+            .store(logical_size as isize, AtomicOrdering::SeqCst);
+        debug!("calculated logical size the hard way: {}", logical_size);
+        Ok(())
+    }
+
+    /// Retrieve current logical size of the timeline
+    ///
+    /// NOTE: counted incrementally, includes ancestors,
+    pub fn get_current_logical_size(&self) -> usize {
+        let current_logical_size = self.current_logical_size.load(AtomicOrdering::Acquire);
+        match usize::try_from(current_logical_size) {
```

Comment:
If there's a bug in tracking the size, we don't want to throw an error and stop ingesting new WAL because of that. It might make sense to set it to a special value that indicates "broken", though, and recalculate later.

If we start calculating the size lazily in the future, a negative value would be useful to track updates, until the lazy calculation finishes. We'd need another variable or enum to indicate whether the calculation has finished or not though.

Reply:
> we don't want to throw an error and stop ingesting new WAL because of that

Hm, you're right. We can continue ingesting new records after the error. IMO it is still better to emit an error when the erroneous modification happens. It can contain more details, e g type of wal record which lead to the error, maybe a backtrace. With current approach we wont notice until someone calls the api or feedback message is generated (it happens quite often but anyway).

> We'd need another variable or enum to indicate whether the calculation has finished or not though.

Thats true.

Reply:
> > we don't want to throw an error and stop ingesting new WAL because of that
> 
> Hm, you're right. We can continue ingesting new records after the error. IMO it is still better to emit an error when the erroneous modification happens. It can contain more details, e g type of wal record which lead to the error, maybe a backtrace. With current approach we wont notice until someone calls the api or feedback message is generated (it happens quite often but anyway).

True, an error in the log when we first detect the problem would be helpful. If the size goes negative, there's a good chance the fault is not with the WAL record where it goes negative, but that we forgot to increment the value at some earlier record. Still, a log message would help to narrow it down.

Reply:
Sort of related: https://github.com/neondatabase/neon/pull/2126#discussion_r930200004 For physical size the reason was in prometheus wrapper. In general I think it would be nice for physical and logical size to agree on what type to use